module4/ex10_wedding_planner/ex10_wedding_planner.py

```diff
@@ -1,10 +1,75 @@
-class GuestList:
+from collections import namedtuple
+
+Person = namedtuple('Person', ['first', 'last'])
+
+
+class TableFull(Exception):
+    """Raised when """
     pass
 
 
+class GuestList:
+    max_at_table = 10
+
+    def __init__(self):
+        self.persons = {}
+        self.tables = {}
+
+    def __len__(self):
+        return len(self.persons.keys())
+
+    def add_to_table(self, person, table_number):
+        if table_number:
+            if table_number in self.tables.keys():
+                self.tables[table_number].append(person)
+                if len(self.tables[table_number]) > self.max_at_table:
+                    raise TableFull("Table is already full")
+            else:
+                self.tables[table_number] = [person]
+        else:
+            if 'unassigned' in self.tables.keys():
+                self.tables['unassigned'].append(person)
+            else:
+                self.tables['unassigned'] = [person]
+
+    def assign(self, person, table_number):
+        if person in self.persons.keys():
+            self.tables[self.persons.get(person)].remove(person)
+            self.add_to_table(person, table_number)
+        else:
+
+            self.persons[person] = table_number
+            self.add_to_table(person, table_number)
+
+    def table(self, table_number):
+        return self.tables[table_number]
+
+    def unassigned(self):
+        return self.tables['unassigned']
+
+    def free_space(self):
+        new_dict = {key: self.max_at_table - len(self.tables[key]) for key in self.tables}
```

Comment:
You can directly return the value of `new_dict` if you are not using it for other purposes as well. It is also recommended not to use too generic variable names, but more specific ones to improve code readability.